### PR TITLE
add ability to receive certificates during pod restart

### PR DESCRIPTION
**Pull request by delgod**

### CODE REVIEW

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py [hunk 1]
```diff
@@ -449,6 +449,36 @@ def _parse_certificates_from_relation_data(relation_data: dict) -> List[Cert]:
                     )
         return certificates
 
+    def get_certificates_for_common_name(self, common_name: str = None, departed_id: int = None) -> List[Cert]:
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v0/tls_certificates.py:452]: Please add unit tests for this new method.

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py [hunk 2]
```diff
@@ -449,6 +449,36 @@ def _parse_certificates_from_relation_data(relation_data: dict) -> List[Cert]:
...
+        """Loops over all relations and returns list of Cert objects.
+
+        Args:
+            common_name: return certificates for specified the common name.
+            departed_id: don't return certificates from departed relation.
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v0/tls_certificates.py:457]: I'm not sure I get the use for this departed_id.

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py [hunk 3]
```diff
@@ -449,6 +449,36 @@ def _parse_certificates_from_relation_data(relation_data: dict) -> List[Cert]:
...
+
+        Returns:
+            list: List of certificates.
+        """
+        certificates = []
+        relations = self.model.relations[self.relationship_name]
+        for relation in relations:
+            if departed_id is not None and relation.id == departed_id:
```
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v0/tls_certificates.py:465]:
```suggestion
            if relation.id == departed_id:
```

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py (hunk 1 above)
Comment (gruyaume) [inline lib/charms/tls_certificates_interface/v0/tls_certificates.py:452]: Why do we allow `common_name` to be none if the method is called `get_certificates_for_common_name`

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py (hunk 3 above)
Comment (delgod) [reply lib/charms/tls_certificates_interface/v0/tls_certificates.py]: fixed in https://github.com/canonical/tls-certificates-interface/pull/4/commits/79e1543206e3b7a6e79f06203a7ff4adc3c8c076

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py (hunk 2 above)
Comment (delgod) [reply lib/charms/tls_certificates_interface/v0/tls_certificates.py]: improved description in https://github.com/canonical/tls-certificates-interface/pull/4/commits/79e1543206e3b7a6e79f06203a7ff4adc3c8c076

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py (hunk 1 above)
Comment (delgod) [reply lib/charms/tls_certificates_interface/v0/tls_certificates.py]: added in https://github.com/canonical/tls-certificates-interface/pull/4/commits/79e1543206e3b7a6e79f06203a7ff4adc3c8c076

Comment (delgod) [reply lib/charms/tls_certificates_interface/v0/tls_certificates.py]: fixed in https://github.com/canonical/tls-certificates-interface/pull/4/commits/79e1543206e3b7a6e79f06203a7ff4adc3c8c076

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py [hunk 4]
```diff
@@ -449,6 +449,40 @@ def _parse_certificates_from_relation_data(relation_data: dict) -> List[Cert]:
                     )
         return certificates
 
+    def get_certificates_for_common_name(
+        self, common_name: str, departed_id: int = None
+    ) -> List[Cert]:
+        """Loops over all relations and returns list of Cert objects.
+
+        Args:
+            common_name: return certificates for specified the common name.
```
Comment (ghislainbourgeois) [inline lib/charms/tls_certificates_interface/v0/tls_certificates.py:458]:
```suggestion
            common_name: return certificates for the specified common name.
```

File: lib/charms/tls_certificates_interface/v0/tls_certificates.py [hunk 5]
```diff
@@ -449,6 +449,40 @@ def _parse_certificates_from_relation_data(relation_data: dict) -> List[Cert]:
...
+                         should be set if the function is executed on relation departed
+                         event to avoid including data from the departed relation.
+
+        Returns:
+            list: List of certificates.
+        """
+        certificates = []
+        relations = self.model.relations[self.relationship_name]
+        for relation in relations:
+            if relation.id == departed_id:
```
Comment (ghislainbourgeois) [inline lib/charms/tls_certificates_interface/v0/tls_certificates.py:469]: Please add a test for this condition, it seems important according to the doc-string and nothing is validating the functionality.

**Review by ghislainbourgeois — CHANGES_REQUESTED** [review]

File: tests/charms/tls_certificates_interface/v0/test_tls_certificates.py [hunk 6]
```diff
@@ -318,3 +320,17 @@ def test_given_valid_relation_data_when_on_relation_changed_then_certificate_ava
             )
         ]
         patch_emit.assert_has_calls(calls, any_order=True)
+
+        self.charm.framework.model.relations = {"certificates": [event.relation]}
```
Comment (gruyaume) [inline tests/charms/tls_certificates_interface/v0/test_tls_certificates.py:324]: This should be a test of its own, let's not bundle multiple behaviour validation in a single test

**Review by gruyaume — CHANGES_REQUESTED** [review]

**Pull request closed without merging**